---
title: "Peak detection"
author: "Mathieu Tiret & Pascal Milesi"
date: "16/09/2021"
output: pdf_document
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# loading libraries. --------------------------------------
if( !requireNamespace("data.table", quietly = T) )
  install.packages("data.table")
if( !requireNamespace("dplyr", quietly = T) )
  install.packages("dplyr")
if( !requireNamespace("tidyverse", quietly = T) )
  install.packages("tidyverse")
if( !requireNamespace("ggplot2", quietly = T) )
  install.packages("ggplot2")
if( !requireNamespace("qvalue", quietly = T) )
  install.packages("qvalue")
if( !requireNamespace("scales", quietly = T) )
  install.packages("scales")

library(data.table)
library(dplyr)
library(ggplot2)
library(tidyverse)
library(qvalue)
library(scales)

# loading datasets. ---------------------------------------
input.path = "." 
output.path = "."
dir.create(output.path) 
dir.create(paste(output.path,"stats/",sep=""))
dir.create(paste(output.path,"plots/",sep=""))
setwd(input.path)
gwas = fread( file.choose() ) %>% as_tibble() # input dataset containing the statistics to be tested (one column per variable).

head(gwas)

list.sign = read.table( file.choose(), header = T ) # Lists of significant SNPs (one columns per variable); can be NULL if not needed. 

head(list.sign)

# declaring variables. ------------------------------------
vars =  # name or column numbers of the variables to be tested. 
slw.size       = 300 # Size of the sliding of the windows 
smooth.intens  = 0.90 # Smoothing intensity for start - end peak definition
extens         = "lfmm" # Extension for output filename  
type           = "Pval" # Type of statistique to work with (any of "BF" for Bayes Factor,"Z" for Z-scores, "Pval" for p.values)
proba          = 0.95 # 0.95 means top 5% of the statistics to be tested
iter           = 10000 # Number of iterations for the randomization process, should be >1000.
```


## Peak detection

```{r}

dataf2 = 
  gwas %>% 
  mutate( Scaff = Scaffold, Pos = 1:n(), ID = ID ) %>% 
  select( Scaff, Pos, ID, LG, all_of(vars) ) %>%
  filter( !is.na(LG) )

liss    = matrix( NA, nrow(dataf2), length(vars) )
signals = list() 
length(signals) = length(vars)

## helping functions. -------------------------------------
computing.profile = function( y, dens.cut )
{
  profile = y > dens.cut
  profile[is.na(profile)] = 0 
  counter = profile[slw.size:length(profile)] - 
    c(0, profile[1:(length(profile)-slw.size)])
  initial.sum = sum( profile[1:(slw.size-1)] )
  
  return( c( rep(NA, slw.size-1), initial.sum + cumsum(counter) ) )
}

randomization = function( .data, iter, dens.cut )
{
  replicate( 
    iter, 
    .data$y %>% sample %>% 
      computing.profile( dens.cut ) %>% max( na.rm = T )
  )
}

## peak detection. ----------------------------------------
for( var in seq_along(vars) ) ####################
{
  # loading data. -------------------------------
  varname = names(dataf2)[4+var]
  dataf = 
    dataf2 %>% 
    rename( y = !!varname ) %>% 
    mutate(
      y = case_when( type == "BF" | type == "Z" ~ log10(y),
                     type == "Pval" ~ -log10(y) )
    )
  
  dens.cut = quantile( dataf$y, p = proba )
  
  cut.quantile = dataf %>% randomization(iter, dens.cut) %>% max
  
  # computing the density. ----------------------
  dataf = 
    dataf %>% 
    group_by(LG) %>% 
    mutate(
      avg    = computing.profile( y, dens.cut ),
      smooth = ifelse( avg < quantile( avg, p = smooth.intens, na.rm = T ),
                       0, avg )
    ) %>% 
    ungroup
  
  # outputting. ---------------------------------
  # smoothed peaks. 
  liss[, var] = dataf %>% pull(smooth)
  
  # positions of the peaks. 
  .signals  = dataf %>% filter( avg > cut.quantile ) %>% pull(Pos)
  .dsignals = diff(.signals)
  
  impl.signals = NULL
  i = 1
  while( i < length(.signals) )
  {
    j = i
    # strictly lower to avoid inter LG peaks. 
    while( j < length(.dsignals) && .dsignals[j] < slw.size ) j = j + 1
    
    impl.signals = 
      c( impl.signals, 
         dataf %>% 
           slice( .signals[i:(j-1)] ) %>% 
           filter( avg == max(avg) ) %>% 
           pull(Pos) %>% median() %>% round()
         )
    i = j + 1
  }
  
  signals[[var]] = impl.signals
  
  ## displaying. ------------------------------------------
  print( paste( round(var/length(vars)*100, 0), " %", sep = "" ) )
}


peak.annotation = list()
cand = list()
cand2 = list()
snp.sign = list()
scaff.sign = list()

for( var in seq_along(vars) )
{
  # computing the peak limits. ------------------
  if(length(signals[[var]]) > 0){
  peaklimit = array(NA, c( length(signals[[var]]), 2 ) )
  for( i in seq_along(signals[[var]]) )
  {
    peaklimit[i, 1] = 
      which( liss[,var] == 0 & 
               seq_along(liss[,var]) < signals[[var]][i] ) %>% max
    peaklimit[i, 2] = 
      which( liss[,var] == 0 & 
               seq_along(liss[,var]) > signals[[var]][i] ) %>% min
  }
  
  # annotating the peaks. -----------------------
  .peak.annotation = matrix( NA, nrow(peaklimit), 8 )
  intervals  = c()
  intervals2 = c()
  .snp.sign = c()
  .scaff.sign = c()
  for( i in 1:nrow(peaklimit) )
  {
    interval = peaklimit[i,1]:peaklimit[i,2]
    tmp.data = dataf2[interval, 4+var] %>% pull
    
    .peak.annotation[i,1] = 
      ifelse( type == "BF" || type == "Z", 
              max( tmp.data ), min( tmp.data ) )
    
    .peak.annotation[i,2] = 
      ifelse(
        type == "BF" || type == "Z",
        interval[ which.max(tmp.data) ],
        interval[ which.min(tmp.data) ] 
      )
    
    .peak.annotation[i,3] = min(interval)
    .peak.annotation[i,4] = max(interval)
    .peak.annotation[i,5] = max(interval) - min(interval)
    
    if(type == "BF" | type == "Z")
    {
      qpeak = which( log10(tmp.data) > 
                       quantile(log10(tmp.data), p = proba) )
    } else 
      qpeak = which( -log10(tmp.data) > 
                       quantile(-log10(tmp.data), p = proba ) )
    
    .peak.annotation[i,6] = length( qpeak )
    .peak.annotation[i,7] = 
      dataf2 %>% 
      slice(c(.peak.annotation[i,3]:.peak.annotation[i,4])) %>% 
      pull(Scaff) %>% unique %>% length 
                              
    .peak.annotation[i,8] = sum( dataf2$ID[ c(.peak.annotation[i,3]:.peak.annotation[i,4]) ] %in% list.sign[,var])
    
    .snp.sign = c(.snp.sign, 
                 dataf2 %>%
                   slice(.peak.annotation[i,3]:.peak.annotation[i,4]) %>%
                   filter(ID %in% list.sign[,var] ) %>% pull(ID))
    
    .scaff.sign = c( .scaff.sign, 
                     dataf2 %>%
                       slice(.peak.annotation[i,3]:.peak.annotation[i,4]) %>% 
                       filter(ID %in% list.sign[,var] ) %>% pull(Scaff))
    
    intervals = c( intervals, peaklimit %>% t %>% c)
    if( dataf2 %>% slice(interval) %>% filter( ID %in% list.sign[,var] ) %>% nrow() > 0 )
    intervals2 = c(intervals2, peaklimit %>% t %>% c)
  }
  
  colnames( .peak.annotation ) = c("Best", "Pos", "Start", "End", "Wide", "SNPs", "Genes", "Sign" )
  rownames( .peak.annotation ) = 
    dataf2 %>% slice( .peak.annotation[,2] ) %>% pull(ID) 
  
  peak.annotation[[var]] = .peak.annotation
  cand[[var]] = unique(intervals)
  cand2[[var]] = unique(intervals2)
  snp.sign[[var]] = .snp.sign
  scaff.sign[[var]] = .scaff.sign
} else{
  peak.annotation[[var]] = rep(NA,8)
  cand[[var]] = NA
  cand2[[var]] = NA
  snp.sign[[var]] = NA
  scaff.sign[[var]] = NA
 }
}

gg_color_hue = function(n) 
{
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}

n    = length(unique(dataf$LG))
cols = gg_color_hue(n)

if( type == "BF" || type == "Z" )
{
  for( var in seq_along(vars) )
  {
    stats = peak.annotation[[var]]
    if(length(stats)==0)
    {
      # do nothing.
    } else {
      write.table(
        stats,
        paste( output.path,"/stats/peak.", extens, ".", colnames(dataf2)[[4+var]], ".", proba, ".txt", sep = "" ),
        col.names = T, row.names = T, quote = F, sep = "\t" ) 
      
     write.table(
        snp.sign[[var]],
        paste( output.path, "/stats/list.snp.peak.", extens, colnames(dataf2)[4+var], ".", proba, ".txt", sep = "" ),
        col.names = F, row.names = F, quote = F, sep = "\t" ) 
      
     write.table(
        scaff.sign[[var]],
        paste( output.path, "/stats/list.scaff.peak.", extens, colnames(dataf2)[4+var], ".", proba, ".txt", sep = "" ),
        col.names = F, row.names = F, quote = F, sep = "\t" ) 
      
      pdf(
        paste( output.path, "/plots/", extens, colnames(dataf2)[4+var], ".", proba, ".",type,".pdf", sep = "" ),
        width = 10, height = 8 ) 
      layout( matrix(c(1,1,1,2,3,3), nrow = 2, byrow = T ) )
      
      plot( log10(dataf2[,4+var] %>% pull),
            main = colnames(dataf2)[4+var],
            col  = cols[factor(dataf2$LG)],
            ylab = colnames(dataf2)[4+var],
            pch = 20, 
            las  = 1
      )

    if(is.na(sum(peak.annotation[[var]]))==F){
      peakheight = ((max(log10(peak.annotation[[var]][,1]))-min(log10(peak.annotation[[var]][,1])))*0.25) / max(liss[,var], na.rm = T) * 1.5

      lines(
        1:nrow(liss) - slw.size/2,
        liss[,var] * peakheight + min(log10(dataf2[,4+var])), 
        col = "gray35" )

      abline( v = c(peak.annotation[[var]][which(peak.annotation[[var]][,8] > 0),2]), col = ("red"), lty = 2, lwd = 1.5 )
      segments( 
        x0 = peak.annotation[[var]][which(peak.annotation[[var]][,8] == 0),2], 
        y0 = min(log10(dataf2[,4+var])), 
        y1 = log10(peak.annotation[[var]][which(peak.annotation[[var]][,8] == 0),1]), 
        col = "gray45", lty = 4)
      
    }
      pdensity = log10(dataf2[,4+var] %>% pull)
      hist( pdensity, freq = F,
            xlab = colnames(dataf2)[4+var],
            las = 1,
            main = "",
            col = "white", 
            border = "white",
            ylim = c(0, max(density(pdensity)$y)) ) 
      lines( density(pdensity), col = "gray35", lwd = 2 )
      abline( v = proba, lty = 3, col = "red" )
      
      plot(
        log10(sample(dataf2[,4+var] %>% pull,replace=F)),
        col = cols[factor(dataf2$LG)],
        ylab = colnames(dataf2)[4+var],
        main = "random")
      
      dev.off()
    }
  }
  
  
} else {
  
  E = sort(-log10(runif(nrow(dataf2),0,1)))
  
  run<-0
  for( var in seq_along(vars) )
  {
    stats = peak.annotation[[var]]
    
    if( length(stats) == 0)
    {
      # do nothing. 
    } else {
      write.table(
        stats,
        paste( output.path , "/stats/peak.", extens, ".", colnames(dataf2)[[4+var]], ".", proba, ".txt", sep = "" ),
        col.names = T, row.names = T, quote = F, sep = "\t" )
      
      
     write.table(
        snp.sign[[var]],
        paste( output.path , "/stats/list.snp.peak.", extens, colnames(dataf2)[4+var], ".", proba, ".txt", sep = "" ),
        col.names = F, row.names = F, quote = F, sep = "\t" )
      
     write.table(
        scaff.sign[[var]],
        paste( output.path , "stats/list.scaff.peak.", extens, colnames(dataf2)[4+var], ".", proba, ".txt", sep = "" ),
        col.names = F, row.names = F, quote = F, sep = "\t" )
     
      pdf(
        paste( output.path , "/plots/", extens, ".", colnames(dataf2)[4+var], ".", proba, ".qval.pdf", sep = "" ),
        width = 10, height = 8 )
      layout( matrix(c(1,1,1,2,3,3), nrow = 2, byrow = T ) )

    if(length(peak.annotation[[var]][which(peak.annotation[[var]][,8] ==0 )])){
      plot( -log10(dataf2[,4+var] %>% pull),
            main = colnames(dataf2)[4+var],
            col  = cols[factor(dataf2$LG)],
            ylab = "-log10(p.values)",
            pch = 20, 
            las  = 1
      )
       if(is.na(sum(peak.annotation[[var]]))==F){
        peakheight = ((max(-log10(peak.annotation[[var]][,1])) - min(-log10(peak.annotation[[var]][,1])))*0.25) / max(liss[,var], na.rm = T) * 1.5

       lines(
         1:nrow(liss) - slw.size/2,
         liss[,var] * peakheight + min(-log10(dataf2[,4+var])), 
         col = "gray35" 
         )

       abline( v = c(peak.annotation[[var]][which(peak.annotation[[var]][,8] > 0),2]), col = ("red"), lty = 2, lwd = 1.5 )
      
        segments( 
         x0 = peak.annotation[[var]][which(peak.annotation[[var]][,8] == 0),2], 
         y0 = min(-log10(dataf2[,4+var])), 
         y1 = -log10(peak.annotation[[var]][which(peak.annotation[[var]][,8] == 0),1]), 
         col = "gray45", lty = 4)
        }
      }else {
          plot( -log10(dataf2[,4+var] %>% pull),
            main = colnames(dataf2)[4+var],
            col  = cols[factor(dataf2$LG)],
            ylab = "-log10(p.values)",
            pch = 20, 
            las  = 1
      )
      
        if(is.na(sum(peak.annotation[[var]]))==F){
      
      peakheight = ((max(-log10(peak.annotation[[var]][,1]))-min(-log10(peak.annotation[[var]][,1])))*0.25) / max(liss[,var], na.rm = T) * 1.5

      lines(
        1:nrow(liss) - slw.size/2,
        liss[,var] * peakheight + min(-log10(dataf2[,4+var])), 
        col = "gray35" )

      abline( v = c(peak.annotation[[var]][which(peak.annotation[[var]][,8] > 0),2]), col = ("red"), lty = 2, lwd = 1.5 )
        }
      }
      
      
      P = sort(-log10(dataf2[,4+var] %>% pull))
      plot( P~E,
            xlab = "Expected p.values",
            las = 1,
            ylab = "Observed p.values" )
      
      A = seq(0,max(E),0.1)
      points(A, A, type = "l", col = "red" )
      points(median(E),median(E),lwd = 1,cex = 1.5,col="red")
      points(median(-log10(dataf2[,4+var] %>% pull)),median(E),lwd = 1,col="red",pch =3,cex = 1.5)
      
      plot( -log10(sample(dataf2[,4+var] %>% pull,replace=F)),
            col = cols[factor(dataf2$LG)],
            ylab = "-log10(p.values)",
            main = "random"
      )
      dev.off()
    }
  }
}


out2 = matrix(NA,length(vars),2)
for( var in seq_along(vars) ) 
{
  out = c(nrow(peak.annotation[[var]]),
          length(unique(peak.annotation[[var]][which(peak.annotation[[var]][,8] >0),2])))
  out2[var,] = out
}

rownames(out2) = colnames(dataf2)[5:(4+length(vars))]
colnames(out2) = c("Peaks","Peaks.Sign")

write.table(
  out2,
  paste(output.path, "stats/Bilan.peaks.", extens, ".txt", sep = "" ),
  col.names = T, row.names = T, quote = T, sep = "\t" )

```
